### ISSUE TITLE
Support of YCCC SO2R box

### ISSUE THREAD

**Issue by ny4i** [n4af/TR4W#61]

**Comment by ny4i** [comment]
Documentation: http://www.k1xm.org/OTRSP/OTRSP_Protocol.pdf

**Comment by gm0gav** [comment]
Yes, this is a priority. As is Microham SO2R support, both should prpbably be done at same time

**Comment by ny4i** [comment]
Adding a comment from a user to show more interest...

**Comment by ny4i** [comment]
Reference device to test OTRSP commands without an actual device: https://k1xm.org/OTRSP/OTRSP_Reference_Device_Installer.exe

**Comment by ny4i** [comment]
During Kyle AA0Z's presentation there was anther request to support this interface. So bump this up the list to a release in a few months. Testing may be an issue but code it to the spec and see if there is a simulator for development.

**Comment by n4af** [comment]
If it helps, i can send u yccc box...

**Comment by ny4i** [comment]
Let me take a crack at the code first. I’ll let you know when it’s ready to test. You can try and if I don’t get it right, we can go to that step. 

Tom

Principal Solutions Architect
Better Software Solutions, Inc. 
727-437-2771

**Comment by ny4i** [comment]
I learned something new today. My Flex supports OTRSP.

"All Flex 6000 rigs implement the OTRSP protocol "

I have a box to implement this with then. I will add this when I add the networking interface for the Flex next month.

**Comment by ny4i** [comment]
I have the test program running and can send commands to it. I believe I need some help to go from the OTRSP commands documented here https://k1xm.org/OTRSP/OTRSP_Protocol.pdf and know when each command in TR4W should send what command.

**Comment by n4af** [comment]
Hi Tom -

So basically, this
<https://www.k1xm.org/OTRSP/OTRSP_Reference_Device_Installer.exe> runs and
you need interpretation of what actions it triggers in TR4W ?

**Comment by n4af** [comment]
i have asked na8v and n5aw to provide some input.  
Using winkey to do the switching works perfect for me, but d/n provide switching for fone.

The RX switching is between otrsp and hardware (audio input).

TR4W major control is via  SwapRadios in LogSubs1.pas

I will take a closer look today...

**Comment by n4af** [comment]
ok, so some basic checking. if we do not issue MAP, then R1/R2 are defined
by port assignment.

So TX1 command checks if active radio is 2 then SWAPRADIOS
  else
    If ActiveRadio is already 1 then
     exit
------------
The RX commands are important but would only be passed on to the OTRSP box
for physical switching of headphones ?

**Comment by ny4i** [comment]
Howie,

That is the right track. I am thinking I need something such as when we SWAPRADIOS, I send a command to OTRSP to swap the RX and TX.

There is also the question if there are manual controls on the OTRSP box that should change states within TR4W. From looking at the YCCC box, there are switches so I suspect if a switch is pressed, I will get a message from the control program and have to take the correct action.

Maybe we just need a state table.

Tom

**Comment by ny4i** [comment]
One additional observation. Adding support for a brand new keyed may not be trivial just because of how much code has exceptions such as If WinKeyer...else ... It would be better to just put the CW sending code in one class and leave the decisions to each object based on keyer type and simply call functions. It's good to explore this but these changes would be after I add the Flex TCP/IP code and a few other changes.

**Comment by n4af** [comment]
On Tue, Apr 26, 2022 at 10:43 AM Tom Schaefer ***@***.***>
wrote:

> Howie,
>
> That is the right track. I am thinking I need something such as when we
> SWAPRADIOS, I send a command to OTRSP to swap the RX and TX.
>
rr/ either TX1, TX2, or NIL
also commands r1,r2,r1s, and r2s

>
> There is also the question if there are manual controls on the OTRSP box
> that should change states within TR4W. From looking at the YCCC box, there
> are switches so I suspect if a switch is pressed, I will get a message from
> the control program and have to take the correct action.
>
Well, I am thinking the OTRSP box will do most of that standalone. Eg, it
has key line, headphones, and ptt lines that the switches will control...

>
> Maybe we just need a state table.
>
> Tom
>

**Comment by n4af** [comment]
A lot of the commands can be skipped for now.
REQUIRED:
  The *TX1, TX2, RX1, RX2*, *RX1S, RX2S*, RX1R, and RX2R commands. RX1R and
RX2R can be treated as RX1S and RX2S. RTS as a PTT control if appropriate.
I am guessing Mode would be nice, but suspect TR4W can already control that
based on radio physical settings ?

**Comment by ny4i** [comment]
I have a YCCC box in the mail on its way to me. I found a used one so I can implement and test it.

**Comment by n4af** [comment]
WOW, that would be great Tom. I think Na8V dropped tr cuz no otrsp support.

Thanks, Howie

**Comment by ny4i** [comment]
OK. I think this is next. It is not clear yet to me if I should use a socket connection or a serial connection. My YCCC box states the software listens on a socket. But it appears one can open the serial port via USB and talk to it directly.

**Comment by ny4i** [comment]
Almost 10 years in the making...

I started on this patch for those keeping score.
